### PR TITLE
Fix the feature flags in production on hydration

### LINKED ISSUE #4082: Switchable flags don't work in production
## Description

The latest release set the `additional_search_views` to `switchable` in staging and production, and `on` by default.
After deploy in production, this flag is `off`, and cannot be switched on.

## Reproduction

1. Go to https://openverse.org/preferences
2. Check the console
3. See error loggged in the console: `DOMException: Failed to execute 'appendChild' on 'Node': This node type does not support this method.`
4. Also see that the switchable `additional_search_views` flag is not working (you can't see the links in tags and source and creator view links in the single result page). If you manually edit the `features` cookie value, you will see the `additional_search_views` working.

## Additional context

This error is happening during hydration.
The `<li>` item is nested in a `<template>` with a `v-if`. This `<template>` is converted to a `comment` node, and when Vue is trying to add a `li` as a child to the `comment` node, we get an error.
To fix this, we should remove the `template`.

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #4082 by @obulat

## Description

The feature flag store was using the env variable to get the deployment environment and the flag state for this environment. However, the env variable is not available on the client. I moved the setting to `publicRuntimeConfig`. To use the config in the feature flag store functions (`getFlagStatus`, `getFeatureState`), I had to move them inside the store itself. Because `this` was not available in some getters that used these functions, I converted them to actions.

I also updated the e2e setup to only set the feature cookies when we explicitly set some feature state in `preparePageForTests`, otherwise we don't set it to not change the default state.

## Testing Instructions

On main, run the app in production mode using `just frontend/run build` and then `DEPLOYMENT_ENV=production just frontend/run start`.
Go to http://localhost:8443/preferences and open console. You should see the errors mentioned in the linked issue in the console. When you attempt to click on the switchable input, the inner circle disappears.

Do the same in this branch, and you should be able to change the state of the switchable flags without errors. You can check the cookie state to confirm the flag was set. You can also see that additional search views are visible when you turn them on.

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: frontend/test/playwright/playwright.config.ts [hunk 1]
```diff
@@ -28,6 +28,7 @@ const config: PlaywrightTestConfig = {
       API_URL,
       UPDATE_TAPES: UPDATE_TAPES,
       PW: "true",
+      DEPLOYMENT_ENV: "staging",
```
Comment (obulat) [inline frontend/test/playwright/playwright.config.ts:31]:
```suggestion
```

File: frontend/test/playwright/e2e/preferences.spec.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    test(`switching ${name} from ${feature.from} to ${feature.to} saves state in a cookie`, async ({
+      page,
+    }) => {
+      await page.goto(`/preferences`)
+      const featureFlag = await getSwitchableInput(page, name, feature.checked)
+      await featureFlag.click()
+
+      const featureCookie = await getFeatureCookies(page, feature.storageCookie)
+      expect(featureCookie[name]).toEqual(feature.to)
+      expect(1).toEqual(9)
```
Comment (obulat) [inline frontend/test/playwright/e2e/preferences.spec.ts:76]:
```suggestion
```

**Comment by sarayourfriend** [comment]
~Unfortunately I don't see the error on `main` :frowning_face:. I've checked and made sure that I am on the latest commit on main, and don't see anything between d25fa3afa ("(tag: frontend-2024.04.09.03.50.12, tag: api-2024.04.09.03.50.11) Turn on additional search views frontend (#4043)") and HEAD that would influence this.~

~I've also tried enabling and disabling the feature, going back and forth between different pages, and I never get the error described in the issue.~

It turns out I didn't understand these instructions correctly:

> On main, run the app in production mode using just frontend/run build and then just frontend/run start.

I was just running the app using build and start, but did not pass `DEPLOYMENT_ENV=production` to the environment. I'll update the testing instructions to clarify this. I realised this after remembering the environment variable being involved in the code changes.

The error shows up for me now, so I'll get back to testing this.

**Review by sarayourfriend — APPROVED** [review]
This mostly LGTM. I was nervous about this change when I saw the line count, but in reviewing the code, I see that it is mostly moving things around, rather than actually making changes to the logic in any way.

As I mentioned private, I still think it is best _not_ to deploy new features to production in the second half of the week, and I did think this was our agreed approach (which is why we target Tuesdays for releases).

However, I am requesting changes, because one of the tests added in `prefereces.spec.ts` is incorrect. The requested change is either to (a) also fix the checkbox bug in this PR, (b) fix the checkbox bug in another PR that blocks this one, or (c) remove that test from this PR, fix the checkbox bug in a critical a11y PR, and in that PR add the test back.

The test as it stands should not be committed to main, though, no more than any other factually and logically incorrect test should.

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
Sorry, I meant to request changes, no approve.

File: frontend/test/playwright/e2e/preferences.spec.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }) => {
+      await page.goto(`/preferences`)
+      const featureFlag = await getSwitchableInput(page, name, feature.checked)
+      await featureFlag.click()
+
+      const inputCheckedStatus = await page
+        .locator(`#${name}`)
+        .first()
+        .getAttribute("checked")
+      expect(inputCheckedStatus).toEqual(feature.checked ? "checked" : null)
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply frontend/test/playwright/e2e/preferences.spec.ts:64]: Here's the issue for the checkbox bug: https://github.com/WordPress/openverse/iss

[the remark this answers is not in the document]
Comment (obulat) [reply frontend/test/playwright/e2e/preferences.spec.ts]: Updated to use `isChecked` for semantic testing in [076043d](https://github.com/WordPress/openverse/pull/4084/commits/076043d0a00d2c3115b7848244b97ff0cbc73292)

File: frontend/test/playwright/e2e/preferences.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    to: "on",
+    storageCookie: "sessionFeatures",
+  },
+  additional_search_views: {
+    checked: true,
+    from: "on",
+    to: "off",
+    storageCookie: "features",
+  },
+}
```
[the remark this answers is not in the document]
Comment (obulat) [reply frontend/test/playwright/e2e/preferences.spec.ts]: Updated in [670310f](https://github.com/WordPress/openverse/pull/4084/commits/670310fc66d588a7feee3bb77edd7da0f28b6f98)

File: frontend/test/playwright/utils/navigation.ts [hunk 5]
```diff
@@ -380,7 +378,7 @@ export interface CookieMap {
 
 export const setCookies = async (
   context: BrowserContext,
-  cookies: CookieMap
+  cookies: Record<string, unknown>
```
[the remark this answers is not in the document]
Comment (obulat) [reply frontend/test/playwright/utils/navigation.ts]: Change reverted.
There was a type error when I made the `features` cookie optional. Fixed in 
[1b672a0](https://github.com/WordPress/openverse/pull/4084/commits/1b672a02470da4d0607f418eec34c0bfc7871052) by making the `cookiesToSet` to be of type `CookieMap`.

**Review by dhruvkb — APPROVED** [review]
LGTM, the changes work as expected locally. I am able to toggle all the preferences and see their effect in the app as well as in the cookies.

**Review by zackkrida — APPROVED** [review]
I think this looks good :+1: Looking forward to a future approach where we do not need to specify a list of feature flag values in frontend/test/playwright/utils/navigation.ts

**Pull request merged**